### PR TITLE
Save edited state of a message when switching rooms

### PR DESCRIPTION
**Pull request by jaiwanth-v**
Fixes vector-im/element-web/issues/17202

https://user-images.githubusercontent.com/65661241/117564950-72c0ff80-b0cc-11eb-8fb9-dba01c108fa3.mp4

Signed-off-by: Jaiwanth <jaiwanth2011@gmail.com>

### CODE REVIEW

File: src/components/views/rooms/EditMessageComposer.js [hunk 1]
```diff
@@ -174,10 +176,42 @@ export default class EditMessageComposer extends React.Component {
...
+
+    _restoreStoredEditorState(partCreator) {
+        const json = localStorage.getItem(this._editorStateKey);
+        if (json) {
+            try {
+                const {parts: serializedParts} = JSON.parse(json);
+                const parts = serializedParts.map(p => partCreator.deserializePart(p));
+                return parts;
+            } catch (e) {
+                console.error(e);
```
[the remark this answers is not in the document]
Comment (jaiwanth-v) [reply src/components/views/rooms/EditMessageComposer.js:196]: Yes

File: src/components/views/rooms/EditMessageComposer.js [hunk 2]
```diff
@@ -195,13 +229,13 @@ export default class EditMessageComposer extends React.Component {
         const editedEvent = this.props.editState.getEvent();
         const editContent = createEditContent(this.model, editedEvent);
         const newContent = editContent["m.new_content"];
-
```
[the remark this answers is not in the document]
Comment (jaiwanth-v) [reply src/components/views/rooms/EditMessageComposer.js:198]: Sure

File: src/components/structures/MessagePanel.js [hunk 3]
```diff
@@ -564,15 +565,23 @@ export default class MessagePanel extends React.Component {
         return ret;
     }
 
+    _wasEventBeingEdited = (mxEv) => {
+        return localStorage.getItem(`mx_edit_state_${mxEv.getRoomId()}
+        _${mxEv.getId()}`) !== null;
+    }
```
[the remark this answers is not in the document]
Comment (jaiwanth-v) [reply src/components/structures/MessagePanel.js:573]: The key is cleared on sending or cancelling any edit, so we're essentially storing only one field per room.

**Comment by jaiwanth-v** [comment]
Also, should we scroll users to the edited event? The cursor is focused on the edit composer, so it might not be obvious if the message is out of timeline view.

**Review by turt2live — APPROVED** [review]
seems sensible, thanks! I don't think we need to scroll for the user - the scroll position is already supposed to be saved, so that information would typically carry through on most user interactions.

**Pull request merged**